src/picongpu/include/fields/background/templates/bgrTWTS.hpp

```diff
@@ -0,0 +1,953 @@
+
+namespace picongpu
+{
+    /** Load external TWTS field
+     *
+     */
+    namespace templates
+    {
+        namespace pmMath = PMacc::algorithms::math;
+        
+        HINLINE
+        TWTSFieldE::TWTSFieldE( const float_64 focus_y_SI,
+                                const float_64 wavelength_SI,
+                                const float_64 pulselength_SI,
+                                const float_64 w_x_SI,
+                                const float_64 w_y_SI,
+                                const float_X phi,
+                                const float_X beta_0,
+                                const float_64 tdelay_user_SI,
+                                const bool auto_tdelay ) :
+            focus_y_SI(focus_y_SI), wavelength_SI(wavelength_SI),
+            pulselength_SI(pulselength_SI), w_x_SI(w_x_SI),
+            w_y_SI(w_y_SI), phi(phi), beta_0(beta_0),
+            tdelay_user_SI(tdelay_user_SI), dt(SI::DELTA_T_SI),
+            unit_length(UNIT_LENGTH), auto_tdelay(auto_tdelay)
+        {
+            /* Note: These objects cannot be instantiated on CUDA GPU device. Since this is done
+                     on host (see fieldBackground.param), this is no problem. */
+            const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
+            halfSimSize=subGrid.getGlobalDomain().size / 2;
+        }
+        
+        HDINLINE PMacc::math::Vector<float3_X, FieldE::numComponents>
+        TWTSFieldE::dummyCast(const PMacc::math::Vector<float3_X, FieldE::numComponents>& x) const
+        { return x; }
+
+        HDINLINE PMacc::math::Vector<float3_X, FieldE::numComponents>
+        TWTSFieldE::dummyCast(const PMacc::math::Vector<float2_X, FieldE::numComponents>& x) const
+            {
+                PMacc::math::Vector<float3_X, FieldE::numComponents> result;
+
+                for( uint32_t i = 0; i < numComponents; ++i ) // cellIdx Ex, Ey and Ez
+                {
+                    result[i]=float3_X( (x[i]).x(),(x[i]).y(),float_X(0.0) );
+                }
+                return result;
+            }
+        
+        template<>
+        HDINLINE PMacc::math::Vector<float3_64,FieldE::numComponents>
+        TWTSFieldE::getEfieldPositions_SI<DIM3>(const DataSpace<simDim>& cellIdx) const
+        {
+            /* Note: Neither direct precisionCast on picongpu::cellSize
+                     or casting on floatD_ does work. */
+            floatD_X cellDim;
+            for (uint32_t i = 0; i<simDim;++i) cellDim[i]=picongpu::cellSize[i];
+                //const floatD_X cellDim = picongpu::cellSize;
+                const float3_64 cellDimensions = precisionCast<float_64>( dummyCast(cellDim) )
+                                                 * unit_length;
+                
+                /* TWTS laser coordinate origin is centered transversally and defined longitudinally
+                 * by the laser center (usually maximum of intensity) in y. */
```

Comment:
basically, I just want that you add that you implicitly counted the "origin" in cells since it is not mentioned or reflected by the variable name.